Remove req to call step() in training loop

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#63164 Remove req to call step() in training loop**
* #63163 Pass `_allow_empty_param_list` into func opt ctor
* #63162 Simplify data structures, add uniform approximation, fix mem leak

Differential Revision: [D30284616](https://our.internmc.facebook.com/intern/diff/D30284616)

**Overview:**
This removes the need to call `step()` in the training loop to follow a refactor for all overlapping methods proposed by @rohan-varma. 

Given the current `dist.GradBucket` implementation, I had to resort to moving the delayed ZeRO initialization from iteration `i`'s explicit optimizer step (i.e. the `step()` call) to the beginning of the first hook in iteration `i+1`, where `i` is the iteration number on which DDP finalizes its bucketing (so 2 or 3 depending on if `static_graph=False` or `True` respectively). This is because, to my knowledge, there is no way to tell if a bucket is the last bucket or not on iteration `i`. [`dist.GradBucket.is_last()` currently returns if the bucket index is 0.](https://github.com/pytorch/pytorch/blob/ee82e7a14ea55101dfe4d245c4c9413b8313fc4c/torch/csrc/distributed/c10d/comm.hpp#L65-L67) I am not sure if this is intentional since the bucket with index 0 is actually the first one to have its all-reduce hook fire, contradicting the [documentation](https://pytorch.org/docs/master/ddp_comm_hooks.html#torch.distributed.GradBucket.is_last).

Commit:
Remove req to call step() in training loop

[ghstack-poisoned]

andwgu commented on torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:
```diff
@@ -123,6 +124,63 @@ def _save_ddp_bucket_info(
+        overlap_info.status = _OverlapStatus.DDP_HAS_REBUILT_BUCKETS
+
+    if overlap_info.status == _OverlapStatus.DDP_HAS_REBUILT_BUCKETS:
+        if bucket_index == 0 and len(overlap_info.params_per_bucket) > 0:
+            # This corresponds to the first bucket of the backward pass
+            # immediately after all information has been saved, so we
+            # can perform the delayed ZeRO initialization
+            zero._init_zero_for_overlap()
+        else:
+            # Once DDP buckets have been rebuilt but ZeRO has not been
+            # properly initialized yet, save the information needed
+            _save_ddp_bucket_info(bucket, zero)
+            return _HookDirective.RETURN
```
If `dist.GradBucket.is_last()` is changed so that it actually returns if the bucket corresponds to the last hook to fire, then we can change this code block to be as follows:
```
_save_ddp_bucket_info(bucket, zero)
if bucket.is_last():
    zero._init_zero_for_overlap()
return _HookDirective.RETURN
```
This will move the delayed initialization to be on the same iteration on which DDP finalizes its bucketing strategy -- specifically, at the end of the hook for the last bucket. 

Minor: This will also result in `bucket_index` not needing to be an argument to the function anymore.

mrshenli replied:
Would I be correct if I assume the quick fix can be changing the following line to `buckets_.size() - i - 1`?

https://github.com/pytorch/pytorch/blob/07b00fc3249d1c6f842eae80ead63e5e4bce62aa/torch/csrc/distributed/c10d/reducer.cpp#L458

mrshenli replied:
I take the above comment back. I saw the condition on `bucket_index == 0`. It might be better to update the `is_last()` logic in GradBucket.

andwgu replied:
Filed an issue to track this `is_last()` problem: https://github.com/pytorch/pytorch/issues/63210

mrshenli replied:
Thanks, I will grab this as a bootcamp task.

andwgu commented on torch/distributed/optim/zero_redundancy_optimizer.py:
```diff
@@ -1037,17 +1037,10 @@ def step(
-
-            # `step()` does not actually perform any parameter updates and is
-            # only used for bookkeeping when `overlap_with_ddp=True`
+            logging.warning(
+                "`step()` should not be included in the training loop when "
+                "`overlap_with_ddp=True`"
+            )
             return None
```
`step()` will become a no-op just in case people still call it in the training loop. With the warning, they should know to remove it.

Commit:
Update on "Remove req to call step() in training loop"


Differential Revision: [D30284616](https://our.internmc.facebook.com/intern/diff/D30284616)

**Overview:**
This removes the need to call `step()` in the training loop to follow a refactor for all overlapping methods proposed by @rohan-varma. 

Given the current `dist.GradBucket` implementation, I had to resort to moving the delayed ZeRO initialization from iteration `i`'s explicit optimizer step (i.e. the `step()` call) to the beginning of the first hook in iteration `i+1`, where `i` is the iteration number on which DDP finalizes its bucketing (so 2 or 3 depending on if `static_graph=False` or `True` respectively). This is because, to my knowledge, there is no way to tell if a bucket is the last bucket or not on iteration `i`. [`dist.GradBucket.is_last()` currently returns if the bucket index is 0.](https://github.com/pytorch/pytorch/blob/ee82e7a14ea55101dfe4d245c4c9413b8313fc4c/torch/csrc/distributed/c10d/comm.hpp#L65-L67) I am not sure if this is intentional since the bucket with index 0 is actually the first one to have its all-reduce hook fire, contradicting the [documentation](https://pytorch.org/docs/master/ddp_comm_hooks.html#torch.distributed.GradBucket.is_last).

[ghstack-poisoned]

andwgu commented:
@andwgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30284616).

mrshenli commented on torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:
```diff
@@ -123,6 +124,63 @@ def _save_ddp_bucket_info(
+            # This corresponds to the first bucket of the backward pass
+            # immediately after all information has been saved, so we
+            # can perform the delayed ZeRO initialization
+            zero._init_zero_for_overlap()
```
does this branch also need a `return _HookDirective.RETURN`?

andwgu replied:
We would want it to return `_HookDirective.CONTINUE` in that case since it is on the iteration **after** the DDP bucket information is collected, meaning a partial optimizer step should happen for that hook.

However, I agree with your other comment. I believe we do not need the `_HookDirective` at all and can just check the overlap status against `INITIALIZED`. I have changed it to do this.

mrshenli commented on torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py:
```diff
@@ -212,24 +273,13 @@ def hook_with_zero_fn(
-        # Proceed as normal until the DDP buckets have been rebuilt
-        if not ddp_ref()._has_rebuilt_buckets:  # type: ignore[union-attr]
-            assert overlap_info.status == _OverlapStatus.UNINITIALIZED
+        if _hook_with_zero_step_setup(ddp_ref, zero, bucket) == _HookDirective.RETURN:
```
OOC, why do we need a new `_HookDirective` type, instead of directly checking whether the `_OverlapStatus` state is `INITIALIZED`?

Commit:
Update on "Remove req to call step() in training loop"


Differential Revision: [D30284616](https://our.internmc.facebook.com/intern/diff/D30284616)

**Overview:**
This removes the need to call `step()` in the training loop to follow a refactor for all overlapping methods proposed by @rohan-varma. 

Given the current `dist.GradBucket` implementation, I had to resort to moving the delayed ZeRO initialization from iteration `i`'s explicit optimizer step (i.e. the `step()` call) to the beginning of the first hook in iteration `i+1`, where `i` is the iteration number on which DDP finalizes its bucketing (so 2 or 3 depending on if `static_graph=False` or `True` respectively). This is because, to my knowledge, there is no way to tell if a bucket is the last bucket or not on iteration `i`. [`dist.GradBucket.is_last()` currently returns if the bucket index is 0.](https://github.com/pytorch/pytorch/blob/ee82e7a14ea55101dfe4d245c4c9413b8313fc4c/torch/csrc/distributed/c10d/comm.hpp#L65-L67) I am not sure if this is intentional since the bucket with index 0 is actually the first one to have its all-reduce hook fire, contradicting the [documentation](https://pytorch.org/docs/master/ddp_comm_hooks.html#torch.distributed.GradBucket.is_last).

[ghstack-poisoned]

andwgu commented:
@andwgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30284616).

Commit:
Update on "Remove req to call step() in training loop"


Differential Revision: [D30284616](https://our.internmc.facebook.com/intern/diff/D30284616)

**Overview:**
This removes the need to call `step()` in the training loop to follow a refactor for all overlapping methods proposed by @rohan-varma. 

Given the current `dist.GradBucket` implementation, I had to resort to moving the delayed ZeRO initialization from iteration `i`'s explicit optimizer step (i.e. the `step()` call) to the beginning of the first hook in iteration `i+1`, where `i` is the iteration number on which DDP finalizes its bucketing (so 2 or 3 depending on if `static_graph=False` or `True` respectively). This is because, to my knowledge, there is no way to tell if a bucket is the last bucket or not on iteration `i`. [`dist.GradBucket.is_last()` currently returns if the bucket index is 0.](https://github.com/pytorch/pytorch/blob/ee82e7a14ea55101dfe4d245c4c9413b8313fc4c/torch/csrc/distributed/c10d/comm.hpp#L65-L67) I am not sure if this is intentional since the bucket with index 0 is actually the first one to have its all-reduce hook fire, contradicting the [documentation](https://pytorch.org/docs/master/ddp_comm_hooks.html#torch.distributed.GradBucket.is_last).

[ghstack-poisoned]
